### ISSUE TITLE
Allow statically linked Python like PythonCall? 

### ISSUE THREAD

**Issue by marius311** [JuliaPy/pyjulia#496]
I'm not an expert and don't know the internals, but is there a reason PyCall can't do whatever PythonCall does that lets the user use any Python executable, including ones with a statically linked libpython? Is there anything preventing what they're doing to be used here? A probably related question posted here: https://github.com/JuliaPy/PyCall.jl/issues/988

**Comment by oschulz** [comment]
That would be so awesome!

**Comment by mkitti** [comment]
Here is cjdoris's response to marius311 on the topic:
https://discourse.julialang.org/t/ann-pythoncall-and-juliacall/76778/16

> I’ve encountered that issue before in pyjulia but don’t actually know its cause.

> I imagine the difference is in how the packages load libpython. In JuliaCall, we pass ctypes.pythonapi._handle to PythonCall, which is a pointer to an already-open libpython. I assume PyJulia/PyCall opens libpython itself.

Indeed, he's right:
https://docs.python.org/3/library/ctypes.html

> ctypes.pythonapi
> An instance of [PyDLL](https://docs.python.org/3/library/ctypes.html#ctypes.PyDLL) that exposes Python C API functions as attributes. Note that all these functions are assumed to return C int, which is of course not always the truth, so you have to assign the correct restype attribute to use these functions.

```
$ ldd `which python`
	linux-vdso.so.1 (0x00007ffe269cb000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fcf5c547000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fcf5c53f000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fcf5c537000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fcf5c52f000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcf5c447000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcf5c21f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fcf5c917000)

$ `which python`
Python 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:06:46) [GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> ctypes.pythonapi
<PyDLL 'None', handle 7f106618a2e0 at 0x7f10655c2980>
```

**Comment by oschulz** [comment]
@mkitti so PyCall/pyjulia could do that as well?

**Comment by mkitti** [comment]
I think so. We technically just need the pointer.

**Comment by oschulz** [comment]
Oh that would be awesome! I guess Packages like PySr (@MilesCranmer), diffeqpy (@ChrisRackauckas) and so on would profit a lot from that as well.

**Comment by mkitti** [comment]
While I agree that `dlopen("")` is undocumented at the Julia API level, it does correspond to the documented behavior at thr C API level.

The use of `ctypes.pythonapi._handle` is also equally undocumented. The underlying mechanism basically depends on the same behavior.

**Comment by cjdoris** [comment]
Actually `ctypes.pythonapi` is documented to be a `PyDLL` and `PyDLL._handle` is documented to be the system handle - in this case the underscore is not indicating an internal attribute, but is to avoid name clashes with symbols in the DLL.

**Comment by mkitti** [comment]
You're right, I concede the point.

https://docs.python.org/3/library/ctypes.html#ctypes.PyDLL._handle

Also `dlopen("")` does not work on macOS and really should be `dlopen(C_NULL)` which doesn't work. One would have to do

```julia
ccall(:jl_load_dynamic_library, Ptr{Cvoid}, (Ptr{Nothing},UInt32,Cint), C_NULL, RTLD_GLOBAL, Cint(1))
```

That does work.

xref: https://github.com/JuliaLang/julia/issues/22318

**Comment by mkitti** [comment]
On macOS `ctypes.pythonapi._handle` is `0xfffffffffffffffe`.

```python
In [1]: import ctypes

In [2]: ctypes.pythonapi._handle
Out[2]: 18446744073709551614

In [3]: hex(ctypes.pythonapi._handle)
Out[3]: '0xfffffffffffffffe'
```

This is actually the value of `RTLD_DEFAULT`: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dlsym.3.html

> If dlsym() is called with the special handle RTLD_DEFAULT, then all mach-o macho
     o images in the process (except those loaded with dlopen(xxx,
     RTLD_LOCAL)) are searched in the order they were loaded.  This can be a
     costly search and should be avoided.

**Comment by cjdoris** [comment]
🤯

I've never actually tried JuliaCall on Mac. I wonder if it works. I should really set up tests and CI.

That very last sentence may explain why loading in ~100 symbols takes so long in PythonCall (~1sec), one reason why PyCall is much faster to load.

**Comment by mkitti** [comment]
On macOS, you can just `dlopen` the executable. At the moment the timing does not look terrible.

```julia
In [1]: from julia.api import LibJulia

In [2]: api = LibJulia.load()

In [3]: api.init_julia()

In [4]: api.jl_eval_string(b"""
   ...: import REPL;
   ...: term = REPL.Terminals.TTYTerminal("dumb", stdin, stdout, stderr);
   ...: repl = REPL.LineEditREPL(term, true);
   ...: REPL.run_repl(repl);
   ...: """)

julia> python_path = ccall(:_dyld_get_image_name, Cstring, (UInt32,), 0) |> unsafe_string
"~/miniforge3-x86_64/envs/pyjulia_test_x86_64/bin/python3.11"

julia> python_handle = dlopen(python_path)
Ptr{Nothing} @0x000000021ba297e0

julia> Py_IsInitialized = dlsym(python_handle, :Py_IsInitialized)
Ptr{Nothing} @0x0000000104d78020

julia> ccall(Py_IsInitialized, Cint, ())
1

julia> @btime dlsym(python_handle, :Py_IsInitialized)
  253.612 ns (1 allocation: 16 bytes)
Ptr{Nothing} @0x0000000104d78020

julia> @btime dlsym(RTLD_DEFAULT, :Py_IsInitialized)
  270.565 ns (1 allocation: 16 bytes)
Ptr{Nothing} @0x0000000104d78020
```

PyCall.jl does a lot of symbol loading during precompilation. That is also going to make it difficult for using this pointer though and is also why it doesn't work with a statically linked python executable.

My thought is that this could benefit from a lazy symbol loading scheme such as the one I put into GR.jl:
https://github.com/jheinen/GR.jl/blob/db3e5f53738be892b23317d673179a32b0e50910/src/funcptrs.jl#L74-L86

**Comment by cjdoris** [comment]
> My thought is that this could benefit from a lazy symbol loading scheme such as the one I put into GR.jl:
> https://github.com/jheinen/GR.jl/blob/db3e5f53738be892b23317d673179a32b0e50910/src/funcptrs.jl#L74-L86

Yeah thanks, I've got something similar in a branch somewhere....
